# [1373] 2진수 8진수
2진수가 주어졌을 때, 8진수로 변환하는 프로그램을 작성하시오.

첫째 줄에 2진수가 주어진다. 주어지는 수의 길이는 1,000,000을 넘지 않는다.

첫째 줄에 주어진 수를 8진수로 변환하여 출력한다.

## 파이썬 내장 함수 진수 변환


1. Python hex (16진수), oct(8진수), bin(2진수), int(x, 진수)<br>

근데 이 친구들의 출력 형태는 str 형태이므로 16진수의 int 형태로 바꾸려면 int('0xa',16)로 넣어야 한다.

In [4]:
N = print(oct(int(input(), 2))[2:]) 
# int(a,2)는 2진법으로 바꾸겠다는 의미
#[2:]는 저렇게 바꾸면 타입이 앞에 같이 붙여지기 때문에 없애려고 한것

11001
31


# [1212] 8진수 2진수
8진수가 주어졌을 때, 2진수로 변환하는 프로그램을 작성하시오.

첫째 줄에 8진수가 주어진다. 주어지는 수의 길이는 333,334을 넘지 않는다.

첫째 줄에 주어진 수를 2진수로 변환하여 출력한다. 수가 0인 경우를 제외하고는 반드시 1로 시작해야 한다.

In [7]:
N = print(bin(int(input(), 8))[2:])

314
11001100


# [2089] -2진수

-2진법은 부호 없는 2진수로 표현이 된다. 2진법에서는 20, 21, 22, 23이 표현 되지만 -2진법에서는 (-2)0 = 1, (-2)1 = -2, (-2)2 = 4, (-2)3 = -8을 표현한다. 10진수로 1부터 표현하자면 1, 110, 111, 100, 101, 11010, 11011, 11000, 11001 등이다.

10진법의 수를 입력 받아서 -2진수를 출력하는 프로그램을 작성하시오.

첫 줄에 10진법으로 표현된 수 N이 주어진다.

2진법 수를 출력한다.

In [116]:
N = int(input())
num=''
if not N:
    print('0')
    exit()
while N :
    if N % (-2):
        num='1'+num
        N = N // -2 +1
        
    else :
        num = '0' + num
        N = N // -2
print(num)

-13
110111


# [11576] Base Conversion

타임머신을 개발하는 정이는 오랜 노력 끝에 타임머신을 개발하는데 성공하였다. 미래가 궁금한 정이는 자신이 개발한 타임머신을 이용하여 500년 후의 세계로 여행을 떠나게 되었다. 500년 후의 세계에서도 프로그래밍을 하고 싶었던 정이는 백준 사이트에 접속하여 문제를 풀기로 하였다. 그러나 미래세계는 A진법을 사용하고 있었고, B진법을 사용하던 정이는 문제를 풀 수가 없었다. 뛰어난 프로그래머였던 정이는 A진법으로 나타낸 숫자를 B진법으로 변환시켜주는 프로그램을 작성하기로 하였다. 

N진법이란, 한 자리에서 숫자를 표현할 때 쓸 수 있는 숫자의 가짓수가 N이라는 뜻이다. 예를 들어 N은 17일 때 한 자릿수에서 사용할 수 있는 수는 0, 1, 2, ... , 16으로 총 17가지가 된다.

입력의 첫 줄에는 미래세계에서 사용하는 진법 A와 정이가 사용하는 진법 B가 공백을 구분으로 주어진다. A와 B는 모두 2이상 30이하의 자연수다.

입력의 두 번째 줄에는 A진법으로 나타낸 숫자의 자리수의 개수 m(1 ≤ m ≤ 25)이 주어진다. 세 번째 줄에는 A진법을 이루고 있는 숫자 m개가 공백을 구분으로 높은 자릿수부터 차례대로 주어진다. 각 숫자는 0이상 A미만임이 보장된다. 또한 수가 0으로 시작하는 경우는 존재하지 않는다.

A진법으로 나타낸 수를 10진법으로 변환하였을 때의 값은 양의 정수이며 220보다 작다.

입력으로 주어진 A진법으로 나타낸 수를 B진법으로 변환하여 출력한다.

In [ ]:
x,y = map(int, input().split())
z = int(input())
a = list(map(int, input().split()))
ten = 0
res = []
for i in range(z) :
    ten += a[-1] * (x ** i)
    a.pop(-1)
    
while ten :
    res.append(str(ten % y))
    ten

# [1978] 소수 찾기 

주어진 수 N개 중에서 소수가 몇 개인지 찾아서 출력하는 프로그램을 작성하시오.

첫 줄에 수의 개수 N이 주어진다. N은 100이하이다. 다음으로 N개의 수가 주어지는데 수는 1,000 이하의 자연수이다.

주어진 수들 중 소수의 개수를 출력한다.

## 풀이 )

소수는 1과 자기 자신 외 양의 약수가 없는 1보다 큰 자연수를 말한다.
다시 말해 2부터 자기 자신보다 1 작은 수들인 수들로 자기 자신을 나누었을때 나머지가 0이 되면 안된다는 의미이다.

In [23]:
N = int(input())
n = list(map(int,input().split()))

def prime_number(number):
    if number != 1: # 수가 1이 아닌 것들중에
        for f in range(2, number): # 2부터 자기자신-1 까지
            if number % f == 0: # 나누었을때 0이 되면 안돼
                return False
    else:
        return False # 수가 1인건 당연히 안돼
    return True # 나머지가 소수지!
a = [] 
for i in range(len(n)):
    if prime_number(n[i])is True:
        a.append(n[i])
print(len(a))

4
1 3 5 7
3


# [1929] 소수 구하기

M이상 N이하의 소수를 모두 출력하는 프로그램을 작성하시오.

첫째 줄에 자연수 M과 N이 빈 칸을 사이에 두고 주어진다. (1 ≤ M ≤ N ≤ 1,000,000) M이상 N이하의 소수가 하나 이상 있는 입력만 주어진다.

한 줄에 하나씩, 증가하는 순서대로 소수를 출력한다.

In [29]:
# 시간초과
M = int(input())
N = int(input())

def prime_number(number):
    if number != 1: # 수가 1이 아닌 것들중에
        for f in range(2, number): # 2부터 자기자신-1 까지
            if number % f == 0: # 나누었을때 0이 되면 안돼
                return False
    else:
        return False # 수가 1인건 당연히 안돼
    return True # 나머지가 소수지!

for i in range(M,N+1):
    if prime_number(i) is True:
        print(i,sep='|n')

3
16
3
5
7
11
13


# [6588] 골드바흐의 추측

1742년, 독일의 아마추어 수학가 크리스티안 골드바흐는 레온하르트 오일러에게 다음과 같은 추측을 제안하는 편지를 보냈다.

4보다 큰 모든 짝수는 두 홀수 소수의 합으로 나타낼 수 있다.
예를 들어 8은 3 + 5로 나타낼 수 있고, 3과 5는 모두 홀수인 소수이다. 또, 20 = 3 + 17 = 7 + 13, 42 = 5 + 37 = 11 + 31 = 13 + 29 = 19 + 23 이다.

이 추측은 아직도 해결되지 않은 문제이다.

백만 이하의 모든 짝수에 대해서, 이 추측을 검증하는 프로그램을 작성하시오.

입력은 하나 또는 그 이상의 테스트 케이스로 이루어져 있다. 테스트 케이스의 개수는 100,000개를 넘지 않는다.

각 테스트 케이스는 짝수 정수 n 하나로 이루어져 있다. (6 ≤ n ≤ 1000000)

입력의 마지막 줄에는 0이 하나 주어진다.

각 테스트 케이스에 대해서, n = a + b 형태로 출력한다. 이때, a와 b는 홀수 소수이다. 숫자와 연산자는 공백 하나로 구분되어져 있다. 만약, n을 만들 수 있는 방법이 여러 가지라면, b-a가 가장 큰 것을 출력한다. 또, 두 홀수 소수의 합으로 n을 나타낼 수 없는 경우에는 "Goldbach's conjecture is wrong."을 출력한다.



In [ ]:
# 에라토스테네스의 체
# 찾고자 하는 수까지 TRUE로 채운 리스트를 생성한 후 2를 제외한 2의 배수, 3을 제외한
# ㄷ3의 배수, ... sqrt(n)의 배수는 모두 False로 바꾼다.
# 2~n까지의 숫자들 중 True인 숫자들이 소수가 된다.
def prime_list(n):
    seive=[True] * n
    m = int(n ** 0.5)
    for i in range(2, m+1):
        if seive[i] == True:
            for j in range(i+1,n,i):
                serive[j] = False
                
    return [i for i in range(2,n) if seive[i] == True]

def sosu(n):
    li = prime_list(n)
    idx = max([i for i in range(len(li)) if li[i]<=n/2])
    for i in range(idx,-1,-1):
        for j in range(1,len(li)):
            if li[i]+li[j]==n:
                return [li[i],li[j]]
            
            
for _ in range(int(input())):
    n=int(input())
    print(" ".join(map(str,sosu(n))))

In [ ]:
# 선배 답
from sys import stdin

def Goldbach():
    check = [False, False] + [True] * 1000000
    for i in range(2, 1001):
        if check[i] == True:
            for j in range(i + i, 1000001, i):
                check[j] = False
    while True:
        n = int(stdin.readline())
        if n == 0:
            break
        A = 0
        B = n
        for _ in range(1000000):
            if check[A] and check[B]:
                print(f"{n} = {A} + {B}")
                break
            A += 1
            B -= 1
        else:
            print("Goldbach's conjecture is wrong.")

Goldbach()

# [11653] 소인수분해

정수 N이 주어졌을 때, 소인수분해하는 프로그램을 작성하시오.

첫째 줄에 정수 N (1 ≤ N ≤ 10,000,000)이 주어진다.

N의 소인수분해 결과를 한 줄에 하나씩 오름차순으로 출력한다. N이 1인 경우 아무것도 출력하지 않는다.

In [118]:
num =int(input())
while num != 1:
    for i in range(2,num+1):
        if (num%i == 0):
            print(i)
            num = num //i
            break
            

9991
97
103


# [10872] 팩토리얼

0보다 크거나 같은 정수 N이 주어진다. 이때, N!을 출력하는 프로그램을 작성하시오.

첫째 줄에 정수 N(0 ≤ N ≤ 12)가 주어진다.

첫째 줄에 N!을 출력한다.

In [38]:
N = int(input())
def fact(n):
    f = 1
    for i in range(1,n+1):
        f *=i
    return f
print(fact(N))

10
3628800


# [1676] 팩토리얼 0의 개수


N!에서 뒤에서부터 처음 0이 아닌 숫자가 나올 때까지 0의 개수를 구하는 프로그램을 작성하시오,

첫째 줄에 N이 주어진다. (0 ≤ N ≤ 500)

첫째 줄에 구한 0의 개수를 출력한다

In [114]:
N = int(input())
print(N//5 + N//25 + N//125) # 5,25,125 때문에 0이 나온다.

30
7


# [2004] 조합 0의 개수


(n m)의 끝자리 의 개수를 출력하는 프로그램을 작성하시오.

첫째 줄에 정수 m,n  (0<=m<=n<=2,n!=0)이 들어온다.

(n m) 끝자리 0의 개수를 출력

In [109]:
# 시간 초과
n,m = list(map(int, input().split()))

def fact(n):
    f = 1
    for i in range(1,n+1):
        f *=i
    return f

jo = fact(abs(n))//(fact(abs(m))*fact(abs(n-m)))
count = 0
for i in range(len(str(jo))):
    if jo % 10 == 0:
        count += 1
        jo = (jo // 10)
        
print(count)

25 12
2
